sdk/src/instructions/composites/swap.ts

```diff
@@ -0,0 +1,93 @@
+import {
+  AddressUtil,
+  deriveATA,
+  resolveOrCreateATAs,
+  TransactionBuilder,
+  ZERO,
+} from "@orca-so/common-sdk";
+import { Address } from "@project-serum/anchor";
+import { PublicKey } from "@solana/web3.js";
+import invariant from "tiny-invariant";
+import { WhirlpoolContext, WhirlpoolData } from "../..";
+import { PDAUtil, TickArrayUtil } from "../../utils/public";
+import { SwapInput, swapIx } from "../swap-ix";
+
+/**
+ * Parameters to build a swap transaction.
+ *
+ * @param poolAddress - The public key for the Whirlpool to swap on
+ * @param whirlpool - A {@link WhirlpoolData} on-chain data object for the pool
+ * @param input - A quote on the desired tokenIn and tokenOut for this swap. Use {@link swapQuoteWithParams} or other swap quote functions to generate this object.
+ * @param wallet - The wallet that tokens will be withdrawn and deposit into.
+ * @param resolveAta - If true, function will automatically resolve and create token ATA to receive tokens. If false, it will derive the ATA using the wallet key, but will not create the ATAs.
+ */
+export type SwapBuilderParams = {
+  poolAddress: Address;
+  whirlpool: WhirlpoolData;
+  input: SwapInput;
+  wallet: PublicKey;
+  resolveAta: boolean;
+};
+
+export async function swap(
+  ctx: WhirlpoolContext,
+  params: SwapBuilderParams
+): Promise<TransactionBuilder> {
+  const { poolAddress, whirlpool, input, wallet } = params;
+  const addressKey = AddressUtil.toPubKey(poolAddress);
+  invariant(input.amount.gt(ZERO), "swap amount must be more than zero.");
+
+  // Check if all the tick arrays have been initialized.
+  const tickArrayAddresses = [input.tickArray0, input.tickArray1, input.tickArray2];
+  const tickArrays = await ctx.fetcher.listTickArrays(tickArrayAddresses, true);
+  const uninitializedIndices = TickArrayUtil.getUninitializedArrays(tickArrays);
+  if (uninitializedIndices.length > 0) {
+    const uninitializedArrays = uninitializedIndices
+      .map((index) => tickArrayAddresses[index].toBase58())
+      .join(", ");
+    throw new Error(`TickArray addresses - [${uninitializedArrays}] need to be initialized.`);
+  }
+
+  const { amount, aToB } = input;
+  const txBuilder = new TransactionBuilder(ctx.provider.connection, ctx.provider.wallet);
+
+  let tokenOwnerAccountA: PublicKey, tokenOwnerAccountB: PublicKey;
+  if (params.resolveAta) {
+    const [resolvedAtaA, resolvedAtaB] = await resolveOrCreateATAs(
+      ctx.connection,
+      wallet,
+      [
+        { tokenMint: whirlpool.tokenMintA, wrappedSolAmountIn: aToB ? amount : ZERO },
+        { tokenMint: whirlpool.tokenMintB, wrappedSolAmountIn: !aToB ? amount : ZERO },
+      ],
```

Comment:
I think we are ok. For this to be an issue, we'll need amountSpecifiedIsInput = false, amount value does not represent SOL & we are sending SOL in the swap. This doesn't exist in any of the scenarios below:

Token A is SOL, aToB == true
- if amountSpecifiedIsInput is true, amount is A and we are sending SOL => amount
- if amountSpecifiedIsInput is false, amount is B and we are receiving SOL => ZERO

Token A is SOL, aToB == false
- if amountSpecifiedIsInput is true, amount is B and we are receiving SOL => ZERO
- if amountSpecifiedIsInput is false, amount is A and we are sending SOL => amount

Token B is SOL, aToB == true
- if amountSpecifiedIsInput is true, amount is A and we are receiving SOL => ZERO
- if amountSpecifiedIsInput is false, amount is B and we are sending SOL => amount

Token B is SOL, aToB == false
- if amountSpecifiedIsInput is true, amount is B and we are sending SOL => amount
- if amountSpecifiedIsInput is false, amount is A and we are receiving SOL => ZERO


Also, otherAmountThreshold is usually U64::max for amountSpecifiedIsInput == true. So if we do have that case, we'll have an issue since a user cannot have U64::max for their SOL balance.